### ISSUE TITLE
Use of "normative optional" annotation should be more consistent

### ISSUE THREAD

**Issue by gibson042** [tc39/ecma262#3280]
This describes a couple of related issues regarding normative optionality:
1. In [The SharedArrayBuffer Constructor](https://tc39.es/ecma262/multipage/structured-data.html#sec-sharedarraybuffer-constructor), "_Whenever a host does not provide concurrent access to SharedArrayBuffers it may omit the **"SharedArrayBuffer"** property of the global object_" text seems identical to the requirements of [Normative Optional](https://tc39.es/ecma262/multipage/conformance.html#:~:text=implement-,Normative%20Optional) and should be marked as such (supporting, among other benefits, cross-referencing).
2. In [WeakRef Objects](https://tc39.es/ecma262/multipage/managing-memory.html#sec-weak-ref-objects), [WeakRef.prototype.constructor](https://tc39.es/ecma262/multipage/managing-memory.html#sec-weak-ref.prototype.constructor) is marked normative optional, implying that WeakRef is required but the "constructor" property of its prototype object is not. While this _might_ have been the intent of #2089, there was no discussion on the point and it strikes me as rather unlikely.

**Comment by ljharb** [comment]
SAB should *definitely* be marked as normative optional, as that's explicitly what got consensus.

Re 2, i'm pretty certain that was the exact goal for SES-adjacent concerns, that the constructor property must be deniable.

**Comment by bakkot** [comment]
@mhofman We specifically didn't do that in https://github.com/tc39/ecma262/pull/1903 because some hosts (including notably browsers) _do_ expose it somewhere other than the global. Normative optional features are all-or-nothing, so marking the whole feature as normative optional would mean that if the constructor was not on the global then hosts could not expose it any other way, and hence would not accomplish the goal of allowing the actual behavior browsers implement.

Where on Matrix are those conversations happening? Feel free to DM me if you don't want to put the link here.

**Comment by mhofman** [comment]
The discussion is around here: https://matrixlogs.bakkot.com/TC39_Structs_and_Shared_Structs/2024-02-15#L11

And yes, my question was that if the host does not expose the SAB constructor in any other way (wasm), is it supposed to implement the feature at all? It's not observable today, but it may become observable in the future.

Then of course if the feature is implemented and exposed through a host mechanism, there is also the question of whether the constructor should be on the global.

IMO, Normative optional means that from the author's pov, the feature may or may not be present, and that an implementation is spec compliant if it doesn't implement the feature.

**Comment by mhofman** [comment]
My question is, if the language itself introduces a way to discover intrinsics (or just "hidden" ones), is a conformant implementation allowed to skip all SAB intrinsics, if they're not otherwise reachable any other way.

**Comment by ljharb** [comment]
Yes, that would need to be an explicit part of the getIntrinsics proposal, I think - that no intrinsic names may be returned or accepted unless the value is reachable from user code.
